### PR TITLE
Migrate to `cosmic-text`

### LINKED ISSUE #7616: Migrate to `cosmic-text`
## What problem does this solve or what need does it fill?

Our font rendering is subpar. See #2404 for an example of the rendering problems and #1325 / #1017 for some missing features.

## What solution would you like?

Migrate to the well-made, well-supported [`cosmic-text`] crate that is quickly becoming the ecosystem standard.

## What alternative(s) have you considered?

We could try and fix all of these issues in the assorted font crates we depend on.

## Additional context

Iced's migration should be informative: https://github.com/iced-rs/iced/pull/1697

### PR DESCRIPTION
**Pull request by tigregalis**
# Objective

Fixes #7616 

## Solution

- Migrate to `cosmic_text`, simplifying and enhancing Bevy's text rendering implementation 
- Provide APIs that take advantage of the newly available features

---

### Changed

`TextStyle::font` is no longer a `Handle<Font>`, but a `FontRef`. `From<Handle<Font>>` is implemented for `FontRef`, so in your existing code, you can call `.into()` on the `Handle<Font>` to convert it into a `FontRef`.

The definition of the font size has changed with the migration to cosmic text. The behavior is now consistent with other platforms (e.g. the web), where the font size in pixels measures the height of the font (the distance between the top of the highest ascender and the bottom of the lowest descender). Font sizes in your app need to be rescaled to approximately 1.2x smaller; for example, if you were using a font size of 60.0, you should now use a font size of 50.0. 

### Added

You can now query loaded fonts by family and style using `FontQuery`. `From<FontQuery>` is implemented for `FontRef`, so you can call `.into()` on the `FontQuery` to convert it into a `FontRef`.

You can now load system fonts using `TextPipeline::load_system_fonts`. These fonts can then be queried by a `FontQuery`.

Emojis and bidirectional text are now supported, provided fonts have been loaded that support them (for example, via `TextPipeline::load_system_fonts`).

Check out the new `system_fonts` example in the Bevy repo.

## Migration Guide

The definition of the font size has changed with the migration to cosmic text. Font sizes in your app need to be rescaled to approximately 1.2x smaller; for example, if you were using a font size of 60.0, you should now use a font size of 50.0..

`TextStyle::font` is no longer a `Handle<Font>`, but a `FontRef`. In each `TextStyle`, call `.into()` on the `Handle<Font>` to convert it into a `FontRef`.

---

# Notes

## To do

This is a draft, and there's still quite a bit of work to do. I've (liberally) peppered the code with TODOs but here is a collection:

At a high level this is what was done and still (perhaps) needs doing:
- [x] Create a side-by-side implementation
- [x] Remove ab_glyph implementation and dependencies
- [x] Support system fonts and a way of accessing them
- [ ] Improvements:
  - [ ] store buffers on the Text component
  - [ ] correct text alignment (prerequisite for correct editor support) (this is broken on cosmic-text's side: https://github.com/pop-os/cosmic-text/issues/42#issuecomment-1607731931)
  - [x] consider moving `Font::get_outlined_glyph_texture` to `pipeline.rs` as it has nothing to do with `Font`.
  - [ ] rescale font sizes in all examples
- [ ] incomplete implementations (for feature parity with existing `bevy_text`)
  - [ ] mixed font sizes
    - [ ] add font size spans to cosmic text if not already (upstream), see https://github.com/pop-os/cosmic-text/issues/64 + https://github.com/pop-os/cosmic-text/pull/150
    - [ ] restore mixed font size in bevy
  - [ ] `PositionedGlyph::byte_index` for #1319, or equivalent? this field is not used in Bevy internals or Bevy examples, so guidance is required, may be covered by Editor/Cursor
  - [ ] parse and validate font on load? This involves bringing in [https://crates.io/crates/ttf-parser](ttf-parser)) which is what is used by `fontdb` (perhaps they could re-export it)
- [ ] Migrate to cosmic-text 0.9.0 (simplifies the above implementation)

## Future Work

Preferably done now:

- (tidy up) text sections renamed to text spans (since we're making breaking changes already)
- (tidy up) simplify implementation based on upstream changes or issues addressed, when landed:
  - https://github.com/DioxusLabs/taffy/pull/490 (remove `Mutex` used for `TextMeasure` purposes)
  - https://github.com/pop-os/cosmic-text/issues/125 / https://github.com/pop-os/cosmic-text/pull/128 (take ID when loading font)
  - https://github.com/pop-os/cosmic-text/pull/127 (cache buffer / store buffer(s) on Text component)
  - https://github.com/pop-os/cosmic-text/pull/124 (remove own `BidiParagraphs` iterator)
  - https://github.com/pop-os/cosmic-text/issues/130 / https://github.com/pop-os/cosmic-text/pull/131
  - don't reimplement determining locale (no issue/PR yet)
  - https://github.com/pop-os/cosmic-text/issues/70 (simpler buffer size calculation)
- (feature) support specifying the line height

For future PRs, or should any of these be done in this PR?

- (feature) text spans as entities
- (feature) editor support
- (feature) editor widget
- (feature) configurable font fallback, provided https://github.com/pop-os/cosmic-text/issues/126
- (feature) inline widgets, provided https://github.com/pop-os/cosmic-text/issues/80
- (fix) will we ever address window-independent scaling: https://github.com/bevyengine/bevy/issues/5621

## Guidance needed

I've tried to "slip under the radar" as much as possible when it comes to the API, but in some cases user-facing API changes were unavoidable. Please consider and advise how best to design and expose these APIs.

Of particular note is the new `FontQuery`, and `TextPipeline::load_system_fonts` (which probably needs to run in a parallel async task).

We also need to consider re-exports.

As far as implementation goes:

- I'm not so confident on the `Arc`/`Mutex` I've introduced and how best to handle errors there.
- Loading fonts are currently "infallible": `ab_glyph` used to parse the font when it loaded it into memory as a `FontVec`, but `fontdb` (via `cosmic_text`) doesn't (parsing is a separate step, and we'd have to introduce `ttf-parser`), so currently font data is simply a `Vec<u8>`.
- I've no idea how to handle `Reflect`.

I think that about covers it. Have at it.

### CODE REVIEW

File: crates/bevy_text/src/font_atlas_set.rs [hunk 1]
```diff
@@ -21,6 +19,10 @@ pub struct FontAtlasSet {
 pub struct GlyphAtlasInfo {
     pub texture_atlas: Handle<TextureAtlas>,
     pub glyph_index: usize,
+    pub left: i32,
```
Comment (alice-i-cecile) [inline crates/bevy_text/src/font_atlas_set.rs:22]: Doc strings on these fields will make them easier to use correctly.

File: crates/bevy_text/src/font_atlas_set.rs [hunk 2]
```diff
@@ -94,43 +96,61 @@ impl FontAtlasSet {
...
             })
     }
 
     pub fn num_font_atlases(&self) -> usize {
         self.font_atlases.len()
     }
 }
+
+#[derive(Debug, Clone)]
+pub struct PositionedGlyph {
```
Comment (alice-i-cecile) [inline crates/bevy_text/src/font_atlas_set.rs:146]:
```suggestion
/// A glyph of a font, typically representing a single character, positioned in screen space.
pub struct PositionedGlyph {
```

File: crates/bevy_text/src/font_atlas_set.rs [hunk 3]
```diff
@@ -94,43 +96,61 @@ impl FontAtlasSet {
...
+    pub position: Vec2,
+    pub size: Vec2,
```
Comment (alice-i-cecile) [inline crates/bevy_text/src/font_atlas_set.rs:148]:
```suggestion
    /// The width and height of the glyph in logical pixels.
    pub size: Vec2,
```

File: crates/bevy_text/src/font_atlas_set.rs [hunk 4]
```diff
@@ -94,43 +96,61 @@ impl FontAtlasSet {
...
+    pub atlas_info: GlyphAtlasInfo,
```
Comment (alice-i-cecile) [inline crates/bevy_text/src/font_atlas_set.rs:149]: Doc strings for the other fields please <3 I don't have great suggestions, as I'm still trying to understand this code, but feel free to ping me for wording help.

File: crates/bevy_text/src/pipeline.rs [hunk 5]
```diff
@@ -1,41 +1,225 @@
...
+// TODO: cache buffers / store buffers on the entity
+// TODO: reconstruct byte indices
+// TODO: rescale font sizes in all examples
+// TODO: fix any broken examples
+// TODO: solve spans with different font sizes, see https://github.com/pop-os/cosmic-text/issues/64
+// TODO: (future work) split text entities into section entities
+// TODO: (future work) text editing
+// TODO: font validation
+
+// TODO: the only reason we need a mutex is due to TextMeasure
```
Comment (alice-i-cecile) [inline crates/bevy_text/src/pipeline.rs:30]: @TimJentzsch @nicoburns I remember solving similar headaches in taffy. Do either of you have advice here?

File: crates/bevy_text/src/text.rs [hunk 6]
```diff
@@ -6,6 +6,8 @@ use bevy_utils::default;
 use serde::{Deserialize, Serialize};
 
 use crate::{Font, DEFAULT_FONT_HANDLE};
+// TODO: reexport cosmic_text and these types in the prelude
+pub use cosmic_text::{FamilyOwned, Stretch, Style, Weight};
```
Comment (alice-i-cecile) [inline crates/bevy_text/src/text.rs:10]: Re-exporting `Style` as such will create a nasty clash with `bevy_ui`'s style component.

File: crates/bevy_text/src/text.rs [hunk 7]
```diff
@@ -188,11 +180,107 @@ pub enum BreakLineOn {
     AnyCharacter,
 }
 
-impl From<BreakLineOn> for glyph_brush_layout::BuiltInLineBreaker {
-    fn from(val: BreakLineOn) -> Self {
-        match val {
-            BreakLineOn::WordBoundary => glyph_brush_layout::BuiltInLineBreaker::UnicodeLineBreaker,
-            BreakLineOn::AnyCharacter => glyph_brush_layout::BuiltInLineBreaker::AnyCharLineBreaker,
+#[derive(Clone, Debug, Reflect, FromReflect)]
+pub enum FontRef {
```
Comment (alice-i-cecile) [inline crates/bevy_text/src/text.rs:184]: Doc strings here: it's hard to follow what's going on.

File: crates/bevy_text/src/text.rs [hunk 8]
```diff
@@ -188,11 +180,107 @@ pub enum BreakLineOn {
...
+}
+
+impl From<Handle<Font>> for FontRef {
+    fn from(handle: Handle<Font>) -> Self {
+        Self::Asset(handle)
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct FontQuery {
```
Comment (alice-i-cecile) [inline crates/bevy_text/src/text.rs:201]: Definitely needs doc strings. I also think we should really try and avoid the `Query` name if possible, since this doesn't seem to use bevy_ecs::Query, which is going to be a source of a lot of confusion.

File: examples/ui/system_fonts.rs [hunk 9]
```diff
@@ -0,0 +1,117 @@
+//! This example demonstrates using system fonts.
```
Comment (alice-i-cecile) [inline examples/ui/system_fonts.rs:1]:
```suggestion
//! This example demonstrates using system fonts, which are already installed on the user's computer.
//! 
//! System fonts won't always be present: if they're not found, a fallback will be used instead.
//! If you need consistent font rendering for aesthetic reasons, you should package and ship your own font as an asset instead.
```

File: crates/bevy_text/src/text.rs (hunk 7 above)
Comment (alice-i-cecile) [reply crates/bevy_text/src/text.rs:184]:
```suggestion
/// Identifies a font to use, which is either stored as an [`Asset`] or loaded directly from the user's system.
pub enum FontRef {
```

File: crates/bevy_text/src/text.rs (hunk 8 above)
Comment (alice-i-cecile) [reply crates/bevy_text/src/text.rs:201]: Looking at this more, I think that "system font" is probably a better way to describe this functionality?

File: crates/bevy_text/src/font_atlas.rs [hunk 10]
```diff
@@ -8,40 +7,9 @@ use bevy_render::{
...
-impl From<Point> for SubpixelOffset {
-    fn from(_: Point) -> Self {
-        Self
-    }
-}
-
 pub struct FontAtlas {
     pub dynamic_texture_atlas_builder: DynamicTextureAtlasBuilder,
-    pub glyph_to_atlas_index: HashMap<(GlyphId, SubpixelOffset), usize>,
+    pub glyph_to_atlas_index: HashMap<cosmic_text::CacheKey, (usize, i32, i32, u32, u32)>,
```
Comment (alice-i-cecile) [inline crates/bevy_text/src/font_atlas.rs:12]: I would really like a newtype here rather than the complex tuple type. Those are really easy to screw up.

File: crates/bevy_text/src/text.rs (hunk 8 above)
Comment (tigregalis) [reply crates/bevy_text/src/text.rs:201]: `SystemFont` doesn't quite capture the idea as you can actually use a `FontQuery` to refer to a font you loaded as an asset (since it's always loaded into the `Database`), which I've confirmed works (but I probably need to test for edge cases). `FontQuery` better captures the idea, and it's [the terminology used by `fontdb`](https://docs.rs/fontdb/latest/fontdb/struct.Query.html), but `FontSearch` could maybe work.

File: crates/bevy_text/src/text.rs (hunk 6 above)
Comment (tigregalis) [reply crates/bevy_text/src/text.rs:10]: Would this work?
```rs
pub use cosmic_text::{
    FamilyOwned as FontFamily,
    Stretch as FontStretch,
    Style as FontStyle,
    Weight as FontWeight
};
```

File: crates/bevy_text/src/pipeline.rs (hunk 5 above)
Comment (tigregalis) [reply crates/bevy_text/src/pipeline.rs:30]: @nicoburns referenced [in discord](https://discord.com/channels/691052431525675048/743663673393938453/1114175469466693644) this PR https://github.com/DioxusLabs/taffy/pull/490 which could in theory remove the need for mutexes.

Comment (nicoburns) [reply crates/bevy_text/src/pipeline.rs:29]: Yes, duplicating my comment from Discord to save people clicking through: It potentially helps in two ways:
- It makes the `Send+Sync` bound on measure functions optional (
- It allows measure functions to (mutably if need be) borrow from a global context which is passed in to the compute_layout function

I think it is mostly the latter that will help here, as allows the FontSystem to be stored outside of Taffy and borrowed by measure functions (via their new context parameter).

Comment (nicoburns) [reply crates/bevy_text/src/pipeline.rs:29]: #490 has landed in Taffy main (using a slightly different approach: a value of a user-defined context type is now stored on each node, and a single measure function is passed into `compute_layout` (which can borrow external variables as needed, and only lives as long as a single layout run avoiding most lifetime issues).

**Comment by TotalKrill** [comment]
was this abandoned, and in that case, how would anyone pick this up?

**Comment by TotalKrill** [comment]
Okay so basically, I took and did the quickest way I could figure out, and that was to just squash all the commits by @tigregalis and then merge that one large commit into a new branch, and then do a PR. 

I probably messed something up, but at least I got the `examples/ui/text.rs` to run, and show similar text. 

You can see that PR here: https://github.com/bevyengine/bevy/pull/10193

**Comment by alice-i-cecile** [comment]
Adopted in #10193. Please follow (and chime in) there!
